File: Java/Core/SfRect.java
package com.appbuilders.quienesestepoke.libraries.Surface.Core;

public class SfRect extends Object {

    public float x;
    public float y;
    public float width;
    public float height;

    public SfRect() {
        this.x = 0;
        this.y = 0;
        this.width = 0;
        this.height = 0;
    }

    public SfRect(float x, float  y, float width, float height) {
        this.setRect(x, y, width, height);
    }

    public void setRectEmpty() {
        this.x = 0;
        this.y = 0;
        this.width = 0;
        this.height = 0;
    }

    public void setRect(float x, float  y, float width, float height) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    public void moveRect(float dx, float dy) {
        this.x += dx;
        this.y += dy;
    }

    public void offsetRect(float dw, float dh) {
        this.width += dw;
        this.height += dh;
    }

    public boolean isEmpty() {
        return (this.x == 0 && this.y == 0 && this.width == 0 && this.height == 0);
    }

    public void unionRec(SfRect rect) {
        if (rect != null) {
            this.width = (rect.x + rect.width) - this.x;
            this.height = (rect.y + rect.height) - this.y;
        }
    }

    public void intersectRect(SfRect rect) {
        if (rect != null) {
            float x2, y2;
            x2 = this.x + this.width;
            y2 = this.y + this.height;
            this.x = rect.x;
            this.y = rect.y;
            this.width = x2 - this.x;
            this.height = y2 - this.y;
        }
    }

    public void subtractRect(SfRect rect) {
        if (rect != null) {
            this.width = rect.x - this.x;
            this.height = rect.y - this.y;
        }
    }

    public boolean pointInRect(SfPoint point) {
        boolean ret = false;
        if (point != null) {
            ret = (point.x >= this.x && point.y >= this.y && point.x <= (this.x + this.width) && point.y <= (this.y + this.height));
        }
        return ret;
    }
}

File: Swift/SurfaceEngineDemo/SurfaceEngineDemo/SfRect.swift
//
//  SfRect.swift
//  SurfaceEngine
//
//  Created by Erick Sanchez on 11/10/17.
//  Copyright © 2017 App Builders. All rights reserved.
//

import Foundation

public class SfRect {
    
    var x: Float!;
    var y: Float!;
    var width: Float!;
    var height: Float!;
    
    init() {
        
        self.x = 0;
        self.y = 0;
        self.width = 0;
        self.height = 0;
    }
    
    convenience init(x:Float, y:Float, width:Float, height:Float) {
        
        self.init();
        self.setRect(x: x, y: y, width: width, height: height);
    }
    
    func setRect(x:Float, y:Float, width:Float, height:Float) -> Void {
        
        self.x = x;
        self.y = y;
        self.width = width;
        self.height = height;
    }
    
    func setRectEmpty() -> Void {
        
        self.x = 0;
        self.y = 0;
        self.width = 0;
        self.height = 0;
    }
    
    func moveRect(dx:Float, dy:Float) -> Void {
        
        self.x = self.x + dx;
        self.y = self.y + dy;
    }
    
    func offsetRect(_ dw:Float, dh:Float) ->  Void {
        
        self.width = dw;
        self.height = dh;
    }
    
    func isEmpty() -> Bool {
        
        if ( self.x == nil && self.y == nil && self.width == nil && self.height == nil ) {
            return true;
        }
        
        return false;
    }
    
    func unionRect(_ rect:SfRect!) {
        
        if ( rect != nil ) {
            
            let x2:Float! = rect.x + rect.width;
            let y2:Float! = rect.y + rect.height;
            
            self.width = self.x - x2;
            self.height = self.y - y2;
            
        }
    }
    
    func intersectRect(_ rect:SfRect!) {
        
        if ( rect != nil ) {
            
            let x:Float! = rect.x;
            let y:Float! = rect.y;
            
            let x2:Float! = ( self.x + self.width ) - x;
            let y2:Float! = ( self.y + self.height ) - y;
            
            self.setRect(x: x, y: y, width: x2, height: y2);
        }
    }
    
    func subtractRect(_ rect:SfRect!) {
        
        if ( rect != nil ) {
            
            self.width = rect.x - self.x;
            self.height = rect.y - self.y;
        }
    }
    
    func pointInRect(_ point:SfPoint!) -> Bool {
        
        var ret:Bool! = false;
        if ( point != nil ) {
            ret = ( point.x >= self.x && point.y >= self.y && point.x <= (self.x + self.width) && point.y <= (self.y + self.height) );
        }
        return ret;
    }
}
